File: MyLife/packages/meerkat-native-transport/android/src/main/java/com/mylife/meerkat/nativetransport/MeerkatBleWakeModule.kt
// UNVERIFIED - pending dev build (Plan 42 WP-42C). This Kotlin source is AUTHORED
// in-repo but has NOT been compiled or device-proven: this environment has no
// Android toolchain. Nothing here is evidence for AC-42.3; the JS bridge keeps
// the BLE-wake rung honestly unavailable until a real native module loads (WP-42B).
//
// Android BLE WAKE-ONLY module over android.bluetooth: a BluetoothLeAdvertiser +
// a GATT server exposing a single notify/read wake characteristic, and a
// BluetoothLeScanner + GATT client that reads only that characteristic. Fulfills
// the raw native contract in src/native-types.ts (RawNativeBleWakeModule) under
// BLE_WAKE_NATIVE_MODULE_NAME ('MeerkatBleWake').
//
// NC-42.6 (load-bearing): BLE carries the bounded wake payload and NOTHING ELSE.
// The GATT server publishes ONE read/notify-only characteristic; there is no
// writable/data characteristic. The scanner connects only long enough to read
// that one characteristic, surfaces its bytes as a 'wake' event, and disconnects.
// A value over the size cap is DROPPED before any 'wake' event is emitted.

package com.mylife.meerkat.nativetransport

import android.bluetooth.BluetoothGatt
import android.bluetooth.BluetoothGattCallback
import android.bluetooth.BluetoothGattCharacteristic
import android.bluetooth.BluetoothGattServer
import android.bluetooth.BluetoothGattServerCallback
import android.bluetooth.BluetoothGattService
import android.bluetooth.BluetoothManager
import android.bluetooth.BluetoothProfile
import android.bluetooth.le.AdvertiseCallback
import android.bluetooth.le.AdvertiseData
import android.bluetooth.le.AdvertiseSettings
import android.bluetooth.le.BluetoothLeAdvertiser
import android.bluetooth.le.BluetoothLeScanner
import android.bluetooth.le.ScanCallback
import android.bluetooth.le.ScanFilter
import android.bluetooth.le.ScanResult
import android.bluetooth.le.ScanSettings
import android.content.Context
import android.os.ParcelUuid
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.util.UUID

private class BleCodedException(code: String) : CodedException(code, "Meerkat BLE wake error: $code", null)

private object MeerkatBle {
  // Owned Meerkat BLE UUIDs (mirror the iOS module's constants).
  val SERVICE_UUID: UUID = UUID.fromString("6d796c69-6665-7761-6b65-000000000042")
  val WAKE_CHARACTERISTIC_UUID: UUID = UUID.fromString("6d796c69-6665-7761-6b65-000000000043")
  // NC-42.6 size gate; the app-side codec encodes three tiny fields.
  const val MAX_WAKE_PAYLOAD_BYTES = 512
}

class MeerkatBleWakeModule : Module() {

  private val context: Context
    get() = appContext.reactContext ?: throw BleCodedException("ERR_NO_CONTEXT")

  private val bluetoothManager: BluetoothManager
    get() = context.getSystemService(Context.BLUETOOTH_SERVICE) as? BluetoothManager
      ?: throw BleCodedException("ERR_BLUETOOTH_UNAVAILABLE")

  // Peripheral (advertiser + GATT server) state.
  private var advertiser: BluetoothLeAdvertiser? = null
  private var gattServer: BluetoothGattServer? = null
  private var wakeCharacteristic: BluetoothGattCharacteristic? = null
  @Volatile private var currentPayload: ByteArray? = null
  private var advertiseCallback: AdvertiseCallback? = null

  // Central (scanner + GATT client) state.
  private var scanner: BluetoothLeScanner? = null
  private var scanCallback: ScanCallback? = null

  override fun definition() = ModuleDefinition {
    Name("MeerkatBleWake")

    Events("wake")

    OnDestroy { teardown() }
    OnActivityEntersBackground { teardown() }

    AsyncFunction("advertise") { input: Map<String, Any?> ->
      val payload = input["payload"] as? ByteArray ?: throw BleCodedException("ERR_INVALID_ARGS")
      // NC-42.6 size gate up front.
      if (payload.size > MeerkatBle.MAX_WAKE_PAYLOAD_BYTES) throw BleCodedException("ERR_PAYLOAD_TOO_LARGE")
      startAdvertising(payload)
    }

    AsyncFunction("stopAdvertising") { stopAdvertising() }
    AsyncFunction("scan") { startScanning() }
    AsyncFunction("stopScanning") { stopScanning() }
    AsyncFunction("destroy") { teardown() }
  }

  // MARK: - Peripheral: GATT server with ONE read/notify wake characteristic

  private fun startAdvertising(payload: ByteArray) {
    currentPayload = payload
    val adapter = bluetoothManager.adapter ?: throw BleCodedException("ERR_BLUETOOTH_UNAVAILABLE")
    val adv = adapter.bluetoothLeAdvertiser ?: throw BleCodedException("ERR_BLUETOOTH_UNAVAILABLE")
    advertiser = adv

    // Bring up the GATT server exposing exactly one read/notify characteristic.
    if (gattServer == null) {
      val server = bluetoothManager.openGattServer(context, serverCallback)
      val service = BluetoothGattService(MeerkatBle.SERVICE_UUID, BluetoothGattService.SERVICE_TYPE_PRIMARY)
      val characteristic = BluetoothGattCharacteristic(
        MeerkatBle.WAKE_CHARACTERISTIC_UUID,
        BluetoothGattCharacteristic.PROPERTY_READ or BluetoothGattCharacteristic.PROPERTY_NOTIFY,
        BluetoothGattCharacteristic.PERMISSION_READ, // read-only; no write permission
      )
      service.addCharacteristic(characteristic)
      server.addService(service)
      gattServer = server
      wakeCharacteristic = characteristic
    }

    val settings = AdvertiseSettings.Builder()
      .setAdvertiseMode(AdvertiseSettings.ADVERTISE_MODE_LOW_LATENCY)
      .setConnectable(true)
      .setTimeout(0)
      .build()
    // Advertise the service UUID only; the payload is served over the GATT read,
    // keeping the advertisement itself within the 31-byte legacy budget.
    val data = AdvertiseData.Builder()
      .addServiceUuid(ParcelUuid(MeerkatBle.SERVICE_UUID))
      .build()
    val cb = object : AdvertiseCallback() {}
    advertiseCallback = cb
    adv.startAdvertising(settings, data, cb)
  }

  private fun stopAdvertising() {
    advertiseCallback?.let { advertiser?.stopAdvertising(it) }
    advertiseCallback = null
    gattServer?.close()
    gattServer = null
    wakeCharacteristic = null
    currentPayload = null
  }

  private val serverCallback = object : BluetoothGattServerCallback() {
    override fun onCharacteristicReadRequest(
      device: android.bluetooth.BluetoothDevice?,
      requestId: Int,
      offset: Int,
      characteristic: BluetoothGattCharacteristic?,
    ) {
      val payload = currentPayload
      if (characteristic?.uuid == MeerkatBle.WAKE_CHARACTERISTIC_UUID && payload != null) {
        val value = if (offset in 0..payload.size) payload.copyOfRange(offset, payload.size) else ByteArray(0)
        gattServer?.sendResponse(device, requestId, BluetoothGatt.GATT_SUCCESS, offset, value)
      } else {
        gattServer?.sendResponse(device, requestId, BluetoothGatt.GATT_FAILURE, offset, null)
      }
    }

    // Explicitly REFUSE writes: wake-only, no inbound data path (NC-42.6).
    override fun onCharacteristicWriteRequest(
      device: android.bluetooth.BluetoothDevice?,
      requestId: Int,
      characteristic: BluetoothGattCharacteristic?,
      preparedWrite: Boolean,
      responseNeeded: Boolean,
      offset: Int,
      value: ByteArray?,
    ) {
      if (responseNeeded) {
        gattServer?.sendResponse(device, requestId, BluetoothGatt.GATT_REQUEST_NOT_SUPPORTED, offset, null)
      }
    }
  }

  // MARK: - Central: scan for the service, read the one characteristic, disconnect

  private fun startScanning() {
    val adapter = bluetoothManager.adapter ?: throw BleCodedException("ERR_BLUETOOTH_UNAVAILABLE")
    val s = adapter.bluetoothLeScanner ?: throw BleCodedException("ERR_BLUETOOTH_UNAVAILABLE")
    scanner = s
    val filters = listOf(
      ScanFilter.Builder().setServiceUuid(ParcelUuid(MeerkatBle.SERVICE_UUID)).build()
    )
    val settings = ScanSettings.Builder()
      .setScanMode(ScanSettings.SCAN_MODE_LOW_LATENCY)
      .build()
    val cb = object : ScanCallback() {
      override fun onScanResult(callbackType: Int, result: ScanResult?) {
        val device = result?.device ?: return
        // Connect only to read the wake characteristic, then disconnect.
        device.connectGatt(context, false, gattClientCallback)
      }
    }
    scanCallback = cb
    s.startScan(filters, settings, cb)
  }

  private fun stopScanning() {
    scanCallback?.let { scanner?.stopScan(it) }
    scanCallback = null
  }

  private val gattClientCallback = object : BluetoothGattCallback() {
    override fun onConnectionStateChange(gatt: BluetoothGatt?, status: Int, newState: Int) {
      if (newState == BluetoothProfile.STATE_CONNECTED) {
        gatt?.discoverServices()
      } else if (newState == BluetoothProfile.STATE_DISCONNECTED) {
        gatt?.close()
      }
    }

    override fun onServicesDiscovered(gatt: BluetoothGatt?, status: Int) {
      val characteristic = gatt?.getService(MeerkatBle.SERVICE_UUID)
        ?.getCharacteristic(MeerkatBle.WAKE_CHARACTERISTIC_UUID)
      if (characteristic != null) {
        gatt.readCharacteristic(characteristic) // read-only; never write
      } else {
        gatt?.disconnect()
      }
    }

    override fun onCharacteristicRead(
      gatt: BluetoothGatt?,
      characteristic: BluetoothGattCharacteristic?,
      value: ByteArray,
      status: Int,
    ) {
      if (status == BluetoothGatt.GATT_SUCCESS &&
        characteristic?.uuid == MeerkatBle.WAKE_CHARACTERISTIC_UUID &&
        value.size <= MeerkatBle.MAX_WAKE_PAYLOAD_BYTES // NC-42.6 size gate on read
      ) {
        sendEvent("wake", mapOf("payload" to value))
      }
      gatt?.disconnect()
    }
  }

  private fun teardown() {
    stopAdvertising()
    stopScanning()
  }
}

File: MyLife/packages/meerkat-native-transport/ios/MeerkatBleWakeModule.swift
// UNVERIFIED - pending dev build (Plan 42 WP-42C). This Swift source is AUTHORED
// in-repo but has NOT been compiled or device-proven: this environment has no
// iOS toolchain. Nothing here is evidence for AC-42.3; the JS bridge keeps the
// BLE-wake rung honestly unavailable until a real native module loads (WP-42B).
//
// iOS BLE WAKE-ONLY module over CoreBluetooth. Fulfills the raw native contract
// in src/native-types.ts (RawNativeBleWakeModule) under BLE_WAKE_NATIVE_MODULE_
// NAME ('MeerkatBleWake').
//
// NC-42.6 (the load-bearing invariant): BLE carries the bounded wake payload and
// NOTHING ELSE. There is no readable/writable data characteristic, no GATT data
// channel, no L2CAP stream. The peripheral role publishes ONE notify-only
// characteristic whose value is the opaque wake bytes; the central role scans,
// connects only long enough to read that single characteristic, surfaces its
// bytes as a 'wake' event, and disconnects. Any advertisement/characteristic
// value over the size cap is DROPPED before a 'wake' event is emitted.

import ExpoModulesCore
import CoreBluetooth
import Foundation

// MARK: - Owned Meerkat BLE UUIDs (Plan 42 Phase 1: "UUIDs owned by Meerkat")

private enum MeerkatBle {
  // A random, Meerkat-owned 128-bit service UUID. Advertised by the peripheral
  // and filtered for by the central so only Meerkat wake beacons are surfaced.
  static let serviceUUID = CBUUID(string: "6D796C69-6665-7761-6B65-000000000042")
  // The single notify-only wake characteristic. Its value is the opaque wake
  // payload bytes; there is NO other characteristic on this service.
  static let wakeCharacteristicUUID = CBUUID(string: "6D796C69-6665-7761-6B65-000000000043")
  /// Hard upper bound on wake payload bytes (NC-42.6 size gate). The app-side
  /// codec (ble-backend.ts) encodes three tiny fields; anything larger is hostile
  /// or malformed and is dropped before any event.
  static let maxWakePayloadBytes = 512
}

// MARK: - Stable error codes

private enum BleError: String {
  case invalidArgs = "ERR_INVALID_ARGS"
  case payloadTooLarge = "ERR_PAYLOAD_TOO_LARGE"
  case bluetoothUnavailable = "ERR_BLUETOOTH_UNAVAILABLE"
}

private final class CodedBleException: Exception {
  // Exception.code is `open var code: String` and Exception.name is a settable
  // `open lazy var`; keep the coded error under its own property name and
  // follow the base signatures exactly (same fix as MeerkatCallKitModule).
  private let errorCode: BleError
  init(_ errorCode: BleError) {
    self.errorCode = errorCode
    super.init()
    self.name = errorCode.rawValue
  }
  override var code: String { errorCode.rawValue }
  override var reason: String { "Meerkat BLE wake error: \(errorCode.rawValue)" }
}

// MARK: - Module

public final class MeerkatBleWakeModule: Module {

  private let queue = DispatchQueue(label: "com.mylife.meerkat.ble")

  private var peripheralShim: BlePeripheralShim?
  private var centralShim: BleCentralShim?

  public func definition() -> ModuleDefinition {
    Name("MeerkatBleWake")

    Events("wake")

    OnDestroy { self.teardown() }

    OnAppEntersBackground {
      // Background scanning/advertising is intentionally NOT sustained here; the
      // wake role is foreground-driven. Tear down radios cleanly.
      self.teardown()
    }

    AsyncFunction("advertise") { (input: [String: Any]) in
      guard let payload = input["payload"] as? Data else {
        throw CodedBleException(.invalidArgs)
      }
      // NC-42.6 size gate: reject an oversized wake payload up front.
      guard payload.count <= MeerkatBle.maxWakePayloadBytes else {
        throw CodedBleException(.payloadTooLarge)
      }
      self.startAdvertising(payload: payload)
    }

    AsyncFunction("stopAdvertising") {
      self.queue.sync { self.peripheralShim?.stop() }
    }

    AsyncFunction("scan") {
      self.startScanning()
    }

    AsyncFunction("stopScanning") {
      self.queue.sync { self.centralShim?.stop() }
    }

    AsyncFunction("destroy") { self.teardown() }
  }

  private func startAdvertising(payload: Data) {
    queue.sync {
      let shim = peripheralShim ?? BlePeripheralShim(queue: queue)
      peripheralShim = shim
      shim.advertise(payload: payload)
    }
  }

  private func startScanning() {
    queue.sync {
      let shim = centralShim ?? BleCentralShim(queue: queue) { [weak self] bytes in
        // Deliver ONLY the wake payload bytes; no data channel, no other value.
        self?.sendEvent("wake", ["payload": bytes])
      }
      centralShim = shim
      shim.scan()
    }
  }

  private func teardown() {
    queue.sync {
      peripheralShim?.stop()
      peripheralShim = nil
      centralShim?.stop()
      centralShim = nil
    }
  }
}

// MARK: - Peripheral shim (advertises the wake service + notify characteristic)

private final class BlePeripheralShim: NSObject, CBPeripheralManagerDelegate {
  private let queue: DispatchQueue
  private var manager: CBPeripheralManager?
  private var wakeCharacteristic: CBMutableCharacteristic?
  private var pendingPayload: Data?
  private var serviceAdded = false

  init(queue: DispatchQueue) {
    self.queue = queue
    super.init()
    manager = CBPeripheralManager(delegate: self, queue: queue)
  }

  func advertise(payload: Data) {
    pendingPayload = payload
    if manager?.state == .poweredOn { publish() }
  }

  private func publish() {
    guard let manager = manager, let payload = pendingPayload else { return }
    // Rebuild the single notify-only characteristic with the current payload as
    // its stored value. There is NO writable/data characteristic on this service.
    let characteristic = CBMutableCharacteristic(
      type: MeerkatBle.wakeCharacteristicUUID,
      properties: [.read, .notify],
      value: nil, // dynamic value supplied on read
      permissions: [.readable]
    )
    wakeCharacteristic = characteristic
    let service = CBMutableService(type: MeerkatBle.serviceUUID, primary: true)
    service.characteristics = [characteristic]
    if !serviceAdded {
      manager.removeAllServices()
      manager.add(service)
      serviceAdded = true
    }
    manager.startAdvertising([
      CBAdvertisementDataServiceUUIDsKey: [MeerkatBle.serviceUUID]
    ])
  }

  func stop() {
    manager?.stopAdvertising()
    manager?.removeAllServices()
    serviceAdded = false
    pendingPayload = nil
    wakeCharacteristic = nil
  }

  func peripheralManagerDidUpdateState(_ peripheral: CBPeripheralManager) {
    if peripheral.state == .poweredOn { publish() }
  }

  func peripheralManager(_ peripheral: CBPeripheralManager, didReceiveRead request: CBATTRequest) {
    // Serve ONLY the bounded wake payload; nothing else is exposed.
    guard request.characteristic.uuid == MeerkatBle.wakeCharacteristicUUID,
          let payload = pendingPayload else {
      peripheral.respond(to: request, withResult: .attributeNotFound)
      return
    }
    if request.offset > payload.count {
      peripheral.respond(to: request, withResult: .invalidOffset)
      return
    }
    request.value = payload.subdata(in: request.offset..<payload.count)
    peripheral.respond(to: request, withResult: .success)
  }

  // Explicitly REFUSE writes: this is wake-only, there is no inbound data path.
  func peripheralManager(_ peripheral: CBPeripheralManager, didReceiveWrite requests: [CBATTRequest]) {
    for request in requests {
      peripheral.respond(to: request, withResult: .writeNotPermitted)
    }
  }
}

// MARK: - Central shim (scans, reads the single wake characteristic, disconnects)

private final class BleCentralShim: NSObject, CBCentralManagerDelegate, CBPeripheralDelegate {
  private let queue: DispatchQueue
  private let onWake: (Data) -> Void
  private var manager: CBCentralManager?
  private var connecting: Set<CBPeripheral> = []

  init(queue: DispatchQueue, onWake: @escaping (Data) -> Void) {
    self.queue = queue
    self.onWake = onWake
    super.init()
    manager = CBCentralManager(delegate: self, queue: queue)
  }

  func scan() {
    guard manager?.state == .poweredOn else { return } // starts on didUpdateState
    manager?.scanForPeripherals(
      withServices: [MeerkatBle.serviceUUID],
      options: [CBCentralManagerScanOptionAllowDuplicatesKey: false]
    )
  }

  func stop() {
    manager?.stopScan()
    for p in connecting { manager?.cancelPeripheralConnection(p) }
    connecting.removeAll()
  }

  func centralManagerDidUpdateState(_ central: CBCentralManager) {
    if central.state == .poweredOn { scan() }
  }

  func centralManager(_ central: CBCentralManager, didDiscover peripheral: CBPeripheral, advertisementData: [String: Any], rssi RSSI: NSNumber) {
    // Connect just long enough to read the single wake characteristic.
    connecting.insert(peripheral)
    peripheral.delegate = self
    central.connect(peripheral, options: nil)
  }

  func centralManager(_ central: CBCentralManager, didConnect peripheral: CBPeripheral) {
    peripheral.discoverServices([MeerkatBle.serviceUUID])
  }

  func peripheral(_ peripheral: CBPeripheral, didDiscoverServices error: Error?) {
    guard error == nil, let services = peripheral.services else {
      manager?.cancelPeripheralConnection(peripheral); return
    }
    for service in services where service.uuid == MeerkatBle.serviceUUID {
      peripheral.discoverCharacteristics([MeerkatBle.wakeCharacteristicUUID], for: service)
    }
  }

  func peripheral(_ peripheral: CBPeripheral, didDiscoverCharacteristicsFor service: CBService, error: Error?) {
    guard error == nil, let chars = service.characteristics else {
      manager?.cancelPeripheralConnection(peripheral); return
    }
    for c in chars where c.uuid == MeerkatBle.wakeCharacteristicUUID {
      peripheral.readValue(for: c) // read-only; never write
    }
  }

  func peripheral(_ peripheral: CBPeripheral, didUpdateValueFor characteristic: CBCharacteristic, error: Error?) {
    defer { manager?.cancelPeripheralConnection(peripheral) }
    guard error == nil,
          characteristic.uuid == MeerkatBle.wakeCharacteristicUUID,
          let value = characteristic.value else { return }
    // NC-42.6 size gate on the read path too: drop oversized/malformed values
    // before surfacing a wake event.
    guard value.count <= MeerkatBle.maxWakePayloadBytes else { return }
    onWake(value)
  }

  func centralManager(_ central: CBCentralManager, didDisconnectPeripheral peripheral: CBPeripheral, error: Error?) {
    connecting.remove(peripheral)
  }

  func centralManager(_ central: CBCentralManager, didFailToConnect peripheral: CBPeripheral, error: Error?) {
    connecting.remove(peripheral)
  }
}
